File: src/main/java/com/mercateo/jsonschema/schema/JsonProperty.java
package com.mercateo.jsonschema.schema;

import com.google.common.annotations.VisibleForTesting;
import com.mercateo.immutables.DataClass;
import com.mercateo.jsonschema.property.PropertyType;
import com.mercateo.jsonschema.property.PropertyTypeMapper;
import org.immutables.value.Value;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Value.Immutable
@DataClass
public abstract class JsonProperty {

    public abstract String getId();

    public abstract String getName();

    public abstract PropertyType getType();

    public abstract String getRef();

    public abstract String getDefaultValue();

    @Value.Default
    public boolean isRequired() {
        return false;
    }

    public abstract List<JsonProperty> getProperties();

    public abstract SizeConstraints getSizeConstraints();

    public abstract ValueConstraints getValueConstraints();

    public abstract List<String> getAllowedValues();

    public abstract String getPath();
}

File: src/main/java/com/mercateo/jsonschema/schema/JsonProperty.kt
package com.mercateo.jsonschema.schema

import com.mercateo.jsonschema.property.PropertyType

data class JsonProperty(
        val id: String,
        val name: String,
        val type: PropertyType,
        val ref: String?,
        val defaultValue: String?,
        val isRequired: Boolean = false,
        val properties: List<JsonProperty>,
        val sizeConstraints: SizeConstraints,
        val valueConstraints: ValueConstraints,
        val allowedValues: Set<String>,
        val path: String
)

